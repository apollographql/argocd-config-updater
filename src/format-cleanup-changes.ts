import { dirname } from 'node:path';

export interface CleanupChange {
  prNumber: number;
  prTitle: string;
  prURL: string;
  filename: string;
  environment: string;
  closedAt: string | null;
}

export function formatCleanupChanges(changes: CleanupChange[]): string {
  if (!changes.length) return '';

  // Group by environment, then by app directory
  const grouped = new Map<string, Map<string, CleanupChange[]>>();

  for (const change of changes) {
    const appDir = dirname(change.filename);
    const env = change.environment;

    let envGroup = grouped.get(env);
    if (!envGroup) {
      envGroup = new Map();
      grouped.set(env, envGroup);
    }

    let appChanges = envGroup.get(appDir);
    if (!appChanges) {
      appChanges = [];
      envGroup.set(appDir, appChanges);
    }

    appChanges.push(change);
  }

  const lines: string[] = [];
  const sortedEnvs = [...grouped.entries()].sort(([a], [b]) =>
    a.localeCompare(b),
  );

  for (const [environment, appGroups] of sortedEnvs) {
    lines.push(`## ${environment}`);
    lines.push('');

    const sortedApps = [...appGroups.entries()].sort(([a], [b]) =>
      a.localeCompare(b),
    );

    for (const [appDir, appChanges] of sortedApps) {
      lines.push(`- ${appDir}`);

      const sortedChanges = appChanges.sort((a, b) => a.prNumber - b.prNumber);
      for (const change of sortedChanges) {
        const closedDateStr = change.closedAt
          ? ` (closed ${new Date(change.closedAt).toLocaleDateString('en-US', {
              month: 'short',
              day: 'numeric',
              year: 'numeric',
            })})`
          : '';

        lines.push(
          `- PR [#${change.prNumber}](${change.prURL}): ${change.prTitle}${closedDateStr}`,
        );
      }
      lines.push('');
    }
  }

  const header = `# Found ${changes.length} closed PR${changes.length === 1 ? '' : 's'}\n\n`;
  return header + lines.join('\n').trim();
}
