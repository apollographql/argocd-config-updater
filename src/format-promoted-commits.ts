import { PromotionsByTargetEnvironment } from './promotionInfo';

export function formatPromotedCommits(
  promotionsByFileThenEnvironment: Map<string, PromotionsByTargetEnvironment>,
): string {
  return [...promotionsByFileThenEnvironment.entries()]
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([filename, promotionsByTargetEnvironment]) => {
      const fileHeader = `* ${filename}\n`;
      const byEnvironment = [...promotionsByTargetEnvironment.entries()]
        .sort(([a], [b]) => a.localeCompare(b))
        .map(([environment, promotionSetWithDockerImage]) => {
          const { promotionSet, dockerImageRepository } =
            promotionSetWithDockerImage;
          const {
            trimmedRepoURL,
            gitConfigPromotionInfo,
            dockerImagePromotionInfo,
            links,
          } = promotionSet;
          const lines = [`  - ${environment}\n`];
          for (const link of links) {
            if (link.url) {
              lines.push(`    + [${link.text}](${link.url})\n`);
            } else {
              lines.push(`    + ${link.text}\n`);
            }
          }
          let alreadyMentionedGitNoCommits = false;
          if (
            dockerImageRepository &&
            dockerImagePromotionInfo &&
            dockerImagePromotionInfo.type !== 'no-change'
          ) {
            let maybeGitConfigNoCommits = '';
            if (gitConfigPromotionInfo.type === 'no-commits') {
              alreadyMentionedGitNoCommits = true;
              maybeGitConfigNoCommits =
                ' (and the git ref for the Helm chart made a no-op change to match)';
            }
            lines.push(
              `    + Changes to Docker image \`${dockerImageRepository}\`${maybeGitConfigNoCommits}\n`,
              ...(dockerImagePromotionInfo.type === 'no-commits'
                ? ['No changes affect the built Docker image.']
                : dockerImagePromotionInfo.type === 'unknown'
                  ? [
                      `Cannot determine set of changes to the Docker image: ${dockerImagePromotionInfo.message}`,
                    ]
                  : dockerImagePromotionInfo.commitSHAs.map(
                      (commitSHA) => `${trimmedRepoURL}/commit/${commitSHA}`,
                    )
              ).map((line) => `      * ${line}\n`),
            );
          }
          if (
            gitConfigPromotionInfo.type !== 'no-change' &&
            !alreadyMentionedGitNoCommits
          ) {
            lines.push(
              `    + Changes to Helm chart\n`,
              ...(gitConfigPromotionInfo.type === 'no-commits'
                ? // This one shows up when the ref changes even though there are no

                  // new commits. This is something we do to try to make the ref
                  // match the Docker tag, so it actually does happen frequently
                  // (though usually only when the Docker tag is making a
                  // substantive change) so this message might end up being a bit
                  // spammy; we can remove it if it's not helpful.
                  [
                    'The git ref for the Helm chart has changed, but there are no new commits in the range.',
                  ]
                : gitConfigPromotionInfo.type === 'unknown'
                  ? [
                      `Cannot determine set of changes to the Helm chart: ${gitConfigPromotionInfo.message}`,
                    ]
                  : gitConfigPromotionInfo.commitSHAs.map(
                      (commitSHA) => `${trimmedRepoURL}/commit/${commitSHA}`,
                    )
              ).map((line) => `      * ${line}\n`),
            );
          }
          return lines.join('');
        });
      return fileHeader + byEnvironment.join('\n');
    })
    .join('');
}
