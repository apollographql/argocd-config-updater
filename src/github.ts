import * as core from '@actions/core';
import { getOctokit } from '@actions/github';
import { LRUCache } from 'lru-cache';

export interface ResolveRefToSHAOptions {
  repoURL: string;
  ref: string;
}

export interface GetTreeSHAForPathOptions {
  repoURL: string;
  commitSHA: string;
  path: string;
}

export interface CompareCommitsOptions {
  repoURL: string;
  baseCommitSHA: string;
  headCommitSHA: string;
}

export interface CompareCommitsResult {
  commits: {
    commitSHA: string;
    message: string;
    author: string | null;
    commitUrl: string;
  }[];
}

export interface GitHubClient {
  resolveRefToSHA(options: ResolveRefToSHAOptions): Promise<string>;
  getTreeSHAForPath(options: GetTreeSHAForPathOptions): Promise<string | null>;
  compareCommits(
    options: CompareCommitsOptions,
  ): Promise<CompareCommitsResult | null>;
}

interface OwnerAndRepo {
  owner: string;
  repo: string;
}

function parseRepoURL(repoURL: string): OwnerAndRepo {
  const m = repoURL.match(/\bgithub\.com\/([\w.-]+)\/([\w.-]+?)(?:\.git|\/)?$/);
  if (!m) {
    throw Error(`Can only track GitHub repoURLs, not ${repoURL}`);
  }
  return { owner: m[1], repo: m[2] };
}

interface AllTreesForCommit {
  pathToTreeSHA: Map<string, string>;
  // If true, GitHub's response to the recursive tree fetch was truncated, so on
  // cache miss we fall back to the getContent API.
  truncated: boolean;
}
export class OctokitGitHubClient {
  apiCalls = new Map<string, number>();
  constructor(private octokit: ReturnType<typeof getOctokit>) {}

  async compareCommits({
    repoURL,
    baseCommitSHA,
    headCommitSHA,
  }: CompareCommitsOptions): Promise<CompareCommitsResult | null> {
    const baseIsCommitSHA = baseCommitSHA.match(/([0-9a-fA-F]+)$/);
    const headIsCommitSHA = headCommitSHA.match(/([0-9a-fA-F]+)$/);
    if (!baseIsCommitSHA || !headIsCommitSHA) return null;

    const { owner, repo } = parseRepoURL(repoURL);

    // Include '^' to be inclusive of the head commit.
    const basehead = `${baseCommitSHA}...${headCommitSHA}^`;
    this.logAPICall('repos.compareCommits', `${owner}/${repo} ${basehead}`);
    try {
      const result = (
        await this.octokit.rest.repos.compareCommitsWithBasehead({
          owner,
          repo,
          basehead,
        })
      ).data.commits;
      return {
        commits: result.map((commit) => ({
          commitSHA: commit.sha,
          message: commit.commit.message,
          author: commit.author?.name ?? commit.commit.author?.email ?? null,
          commitUrl: commit.html_url,
        })),
      };
    } catch (error: unknown) {
      core.info(`[GH API] Error in compareCommits, ignoring`);
      return null;
    }
  }

  private logAPICall(name: string, description: string): void {
    core.info(`[GH API]${name} ${description}`);
    this.apiCalls.set(name, (this.apiCalls.get(name) ?? 0) + 1);
  }

  // The cache key is JSON-ification of `{ repoURL, commitSHA }`.
  private allTreesForCommitCache = new LRUCache<
    string,
    AllTreesForCommit,
    { repoURL: string; commitSHA: string }
  >({
    max: 1024,
    // We ignore the key itself and treat the context as the key
    // (the key is just the JSON-ification of context).
    fetchMethod: async (_key, _staleValue, { context }) => {
      const { repoURL, commitSHA } = context;
      const { owner, repo } = parseRepoURL(repoURL);
      this.logAPICall('git.getCommit', `${owner} / ${repo} ${commitSHA}`);
      const rootTreeSHA = (
        await this.octokit.rest.git.getCommit({
          owner,
          repo,
          commit_sha: commitSHA,
        })
      ).data.tree.sha;
      this.logAPICall('git.getTree', `${owner} / ${repo} ${rootTreeSHA}`);
      const { tree, truncated } = (
        await this.octokit.rest.git.getTree({
          owner,
          repo,
          tree_sha: rootTreeSHA,
          recursive: 'true',
        })
      ).data;
      const allTreesForCommit: AllTreesForCommit = {
        pathToTreeSHA: new Map(),
        truncated,
      };
      for (const { path, type, sha } of tree) {
        if (
          type === 'tree' &&
          typeof path === 'string' &&
          typeof sha === 'string'
        ) {
          allTreesForCommit.pathToTreeSHA.set(path, sha);
        }
      }
      // Also set the root itself in case we're tracking the root of a repo for
      // some reason.
      allTreesForCommit.pathToTreeSHA.set('', rootTreeSHA);
      return allTreesForCommit;
    },
  });

  async resolveRefToSHA({
    repoURL,
    ref,
  }: ResolveRefToSHAOptions): Promise<string> {
    // If the ref already looks like a SHA, just return it.
    // This does not validate that the commit actually exists in the repo,
    // but in practice the next thing we're going to do is call getTreeSHAForPath
    // with the SHA and that will apply that validation.
    if (ref.match(/^[0-9a-f]{40}$/)) {
      return ref;
    }
    const { owner, repo } = parseRepoURL(repoURL);
    const prNumber = ref.match(/^pr-([0-9]+)$/)?.[1];
    const refParameter = prNumber ? `pull / ${prNumber} / head` : ref;
    this.logAPICall('repos.getCommit', `${owner} / ${repo} ${refParameter}`);
    const sha = (
      await this.octokit.rest.repos.getCommit({
        owner,
        repo,
        ref: refParameter,
        mediaType: {
          format: 'sha',
        },
      })
    ).data as unknown;
    // The TS types don't understand that `mediaType: { format: 'sha' }` turns
    // `.data` into a string, so we have to cast to `unknown` and check
    // ourselves.
    if (typeof sha !== 'string') {
      throw Error('Expected string response');
    }
    return sha;
  }

  async getTreeSHAForPath({
    repoURL,
    commitSHA,
    path,
  }: GetTreeSHAForPathOptions): Promise<string | null> {
    const allTreesForCommit = await this.allTreesForCommitCache.fetch(
      JSON.stringify({ repoURL, commitSHA }),
      { context: { repoURL, commitSHA } },
    );
    if (!allTreesForCommit) {
      // This shouldn't happen: errors should lead to an error being thrown from
      // the previous line, but the fetchMethod always returns an actual item.
      throw Error(`Unexpected missing entry in allTreesForCommitCache`);
    }
    const shaFromCache = allTreesForCommit.pathToTreeSHA.get(path);
    if (shaFromCache) {
      return shaFromCache;
    }
    if (!allTreesForCommit.truncated) {
      // The recursive listing we got from GitHub is complete, so if it doesn't
      // have the tree in question, then the path just doesn't exist (as a tree)
      // at the given commit.
      return null;
    }
    // Hmm, we haven't heard of this tree but our listing was truncated. Fall
    // back to the one-at-a-time API.
    return this.getTreeSHAForPathViaGetContent({ repoURL, commitSHA, path });
  }

  // Fall back to asking the GitHub API for the tree hash directly if our cache
  // was truncated.
  private async getTreeSHAForPathViaGetContent({
    repoURL,
    commitSHA,
    path,
  }: GetTreeSHAForPathOptions): Promise<string | null> {
    const { owner, repo } = parseRepoURL(repoURL);
    this.logAPICall('repos.getContent', `${owner} / ${repo} ${commitSHA}`);
    let data;
    try {
      data = (
        await this.octokit.rest.repos.getContent({
          owner,
          repo,
          ref: commitSHA,
          path,
          mediaType: {
            format: 'object',
          },
        })
      ).data as unknown;
    } catch (e: unknown) {
      // If it looks like "not found" just return null.
      if (typeof e === 'object' && e && 'status' in e && e.status === 404) {
        return null;
      }
      throw e;
    }
    // TS types seem confused here too; this works in practice.
    if (
      !(
        typeof data === 'object' &&
        data !== null &&
        'type' in data &&
        data.type === 'dir' &&
        'sha' in data &&
        typeof data.sha === 'string'
      )
    ) {
      throw Error('response does not appear to be a tree');
    }
    return data.sha;
  }
}

export class CachingGitHubClient {
  constructor(
    private wrapped: GitHubClient,
    dump?: CachingGitHubClientDump,
  ) {
    if (dump) {
      this.getTreeSHAForPathCache.load(dump.treeSHAs);
    }
  }

  private resolveRefToSHACache = new LRUCache<
    string,
    string,
    ResolveRefToSHAOptions
  >({
    max: 1024,
    fetchMethod: async (_key, _staleValue, { context }) => {
      return this.wrapped.resolveRefToSHA(context);
    },
  });

  private getTreeSHAForPathCache = new LRUCache<
    string,
    // LRUCache can't store null, so we box it.
    { boxed: string | null },
    GetTreeSHAForPathOptions
  >({
    max: 1024,
    fetchMethod: async (_key, _staleValue, { context }) => {
      return { boxed: await this.wrapped.getTreeSHAForPath(context) };
    },
  });

  async resolveRefToSHA(options: ResolveRefToSHAOptions): Promise<string> {
    const sha = await this.resolveRefToSHACache.fetch(JSON.stringify(options), {
      context: options,
    });
    if (!sha) {
      throw Error(
        'resolveRefToSHACache.fetch should never resolve without a real SHA',
      );
    }
    return sha;
  }

  async getTreeSHAForPath(
    options: GetTreeSHAForPathOptions,
  ): Promise<string | null> {
    const cached = await this.getTreeSHAForPathCache.fetch(
      JSON.stringify(options),
      {
        context: options,
      },
    );
    if (!cached) {
      throw Error(
        'getTreeSHAForPathCache.fetch should never resolve without a boxed value',
      );
    }
    return cached.boxed;
  }

  async compareCommits(
    options: CompareCommitsOptions,
  ): Promise<CompareCommitsResult | null> {
    return this.wrapped.compareCommits(options);
  }

  dump(): CachingGitHubClientDump {
    // We don't dump resolveRefToSHACache because it is not immutable (it tracks
    // the current commits on main, etc).
    return { treeSHAs: this.getTreeSHAForPathCache.dump() };
  }
}

export interface CachingGitHubClientDump {
  treeSHAs: [
    string,
    LRUCache.Entry<{
      boxed: string | null;
    }>,
  ][];
}

export function isCachingGitHubClientDump(
  dump: unknown,
): dump is CachingGitHubClientDump {
  if (!dump || typeof dump !== 'object') {
    return false;
  }
  if (!('treeSHAs' in dump)) {
    return false;
  }
  const { treeSHAs } = dump;
  if (!Array.isArray(treeSHAs)) {
    return false;
  }

  // XXX we could check the values further if we want to be more anal
  return true;
}
