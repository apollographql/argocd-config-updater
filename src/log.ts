import * as core from '@actions/core';

export class PrefixingLogger {
  private silent = false;

  constructor(private prefix = '') {}

  static silent(): PrefixingLogger {
    const logger = new PrefixingLogger();
    logger.silent = true;
    return logger;
  }

  info(message: string): void {
    if (!this.silent) {
      core.info(this.prefix + message);
    }
  }
  warning(message: string): void {
    if (!this.silent) {
      core.warning(this.prefix + message);
    }
  }
  error(message: string): void {
    if (!this.silent) {
      core.error(this.prefix + message);
    }
  }

  withExtendedPrefix(extension: string): PrefixingLogger {
    const logger = new PrefixingLogger(this.prefix + extension);
    logger.silent = this.silent;
    return logger;
  }
}
