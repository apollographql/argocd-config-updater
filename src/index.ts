import * as core from '@actions/core';
import * as github from '@actions/github';
import * as glob from '@actions/glob';
import { throttling } from '@octokit/plugin-throttling';
import { eachLimit } from 'async';
import { readFile, writeFile } from 'fs/promises';
import {
  ArtifactRegistryDockerRegistryClient,
  CachingDockerRegistryClient,
  DockerRegistryClient,
} from './artifactRegistry';
import {
  CachingGitHubClient,
  CachingGitHubClientDump,
  GitHubClient,
  OctokitGitHubClient,
  isCachingGitHubClientDump,
} from './github';
import { updateDockerTags } from './update-docker-tags';
import { updateGitRefs } from './update-git-refs';
import { updatePromotedValues } from './update-promoted-values';
import { PrefixingLogger } from './log';
import { inspect } from 'util';

/**
 * The main function for the action.
 * @returns {Promise<void>} Resolves when the action is complete.
 */
export async function main(): Promise<void> {
  try {
    const files = core.getInput('files');
    const globber = await glob.create(files);
    const filenames = await globber.glob();

    core.setOutput(
      'suggested-promotion-branch-name',
      `${core.getInput('promotion-target-regexp')}_${core.getInput('files')}`.replaceAll(
        /[^-a-zA-Z0-9._]/g,
        '_',
      ),
    );

    let gitHubClient: GitHubClient | null = null;
    let finalizeGitHubClient: (() => Promise<void>) | null = null;
    if (core.getBooleanInput('update-git-refs')) {
      const githubToken = core.getInput('github-token');
      const octokit = github.getOctokit(
        githubToken,
        {
          throttle: {
            onRateLimit: (retryAfter, options) => {
              core.warning(
                `[RATE LIMIT] Hit GH rate limit for request ${options.method} ${options.url}; retrying after ${retryAfter} seconds`,
              );
              return true;
            },
            onSecondaryRateLimit: (retryAfter, options) => {
              core.warning(
                `[RATE LIMIT] Hit secondary GH rate limit for request ${options.method} ${options.url}; retrying after ${retryAfter} seconds`,
              );
              return true;
            },
          },
        },
        throttling,
      );

      // Log GH rate limit response headers after each response and at the end.
      let lastRateLimitHeaderInfo: string | null = null;
      octokit.hook.after('request', async (response) => {
        const prefix = 'x-ratelimit-';
        const rateLimitHeaders: string[] = [];
        for (const [name, value] of Object.entries(response.headers)) {
          if (name.startsWith(prefix)) {
            rateLimitHeaders.push(`${name.substring(prefix.length)}=${value}`);
          }
        }

        if (rateLimitHeaders.length) {
          const rateLimitHeaderInfo = rateLimitHeaders.join(', ');
          lastRateLimitHeaderInfo = rateLimitHeaderInfo;
          core.info(`[GH Rate Limit Info] ${rateLimitHeaderInfo}`);
        }
      });

      const octokitGitHubClient = new OctokitGitHubClient(octokit);

      const apiCacheFileName = core.getInput('api-cache');
      let initialAPICache: APICache | null = null;
      if (apiCacheFileName) {
        initialAPICache = await maybeReadAPICache(apiCacheFileName);
      }

      const cachingGitHubClient = new CachingGitHubClient(
        octokitGitHubClient,
        initialAPICache?.gitHub,
      );

      gitHubClient = cachingGitHubClient;

      finalizeGitHubClient = async () => {
        for (const [name, count] of octokitGitHubClient.apiCalls) {
          core.info(`Total GH API calls for ${name}: ${count}`);
        }
        if (lastRateLimitHeaderInfo) {
          core.info(`Last GH Rate Limit Info: ${lastRateLimitHeaderInfo}`);
        }
        if (apiCacheFileName) {
          const finalAPICache: APICache = {
            version: 1,
            gitHub: cachingGitHubClient.dump(),
          };
          await writeFile(apiCacheFileName, JSON.stringify(finalAPICache));
        }
      };
    }

    let dockerRegistryClient: DockerRegistryClient | null = null;
    const artifactRegistryRepository = core.getInput(
      'update-docker-tags-for-artifact-registry-repository',
    );
    if (artifactRegistryRepository) {
      dockerRegistryClient = new CachingDockerRegistryClient(
        new ArtifactRegistryDockerRegistryClient(artifactRegistryRepository),
      );
    }

    const parallelism = +core.getInput('parallelism');
    const errors: { filename: string; error: unknown }[] = [];
    await eachLimit(filenames, parallelism, async (filename) => {
      try {
        await processFile(filename, gitHubClient, dockerRegistryClient);
      } catch (error) {
        errors.push({ filename, error });
      }
    });
    if (errors.length) {
      core.setFailed(
        `Errors occurred while processing ${errors.length} file${errors.length > 1 ? 's' : ''}`,
      );
      for (const { filename, error } of errors) {
        core.error(`Error while processing ${filename}: ${inspect(error)}`);
      }
    } else {
      await finalizeGitHubClient?.();
    }
  } catch (error) {
    // Fail the workflow run if an error occurs
    if (error instanceof Error) core.setFailed(error.message);
  }
}

async function processFile(
  filename: string,
  gitHubClient: GitHubClient | null,
  dockerRegistryClient: DockerRegistryClient | null,
): Promise<void> {
  const shortFilename = filename.startsWith(`${process.cwd()}/`)
    ? filename.substring(process.cwd().length + 1)
    : filename;
  const logger = new PrefixingLogger(`[${shortFilename}] `);
  let contents = await readFile(filename, 'utf-8');

  if (dockerRegistryClient) {
    contents = await updateDockerTags(contents, dockerRegistryClient, logger);
  }

  // The git refs depend on the docker tag potentially so we want to update it after the
  // docker tags are updated.
  if (gitHubClient) {
    contents = await updateGitRefs(contents, gitHubClient, logger);
  }

  if (core.getBooleanInput('update-promoted-values')) {
    const promotionTargetRegexp = core.getInput('promotion-target-regexp');
    contents = await updatePromotedValues(
      contents,
      promotionTargetRegexp || null,
      logger,
    );
  }

  await writeFile(filename, contents);
}

interface APICache {
  version: 1;
  gitHub: CachingGitHubClientDump;
}

async function maybeReadAPICache(
  apiCacheFileName: string,
): Promise<APICache | null> {
  let apiCacheText: string;
  try {
    apiCacheText = await readFile(apiCacheFileName, 'utf8');
  } catch (e) {
    core.error(`Error reading cache file ${apiCacheFileName}, ignoring: ${e}`);
    return null;
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(apiCacheText);
  } catch (e) {
    core.error(`Error parsing cache file ${apiCacheFileName}, ignoring: ${e}`);
    return null;
  }

  if (
    !parsed ||
    typeof parsed !== 'object' ||
    !('gitHub' in parsed) ||
    !('version' in parsed) ||
    parsed.version !== 1
  ) {
    core.error(
      `Cache file ${apiCacheFileName} has the wrong structure; ignoring`,
    );
    return null;
  }

  if (!isCachingGitHubClientDump(parsed.gitHub)) {
    core.error(
      `Cache file ${apiCacheFileName} has the wrong structure under 'gitHub'; ignoring`,
    );
    return null;
  }

  return { version: parsed.version, gitHub: parsed.gitHub };
}

main();
