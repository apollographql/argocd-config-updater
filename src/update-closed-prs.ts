import { getWebURL, GitHubClient } from './github';
import { PrefixingLogger } from './log';
import { parseYAML } from './yaml';
import { findTrackables } from './update-git-refs';
import { CleanupChange } from './format-cleanup-changes';

/**
 * Extract app name from file path.
 * Examples:
 *   - "teams/backend/test-app/application-values.yaml" -> "test-app"
 *   - "/path/to/teams/test-api/application-values.yaml" -> "test-api"
 */
function extractAppNameFromFilename(filename: string): string {
  const parts = filename.split('/');
  const appValuesIndex = parts.indexOf('application-values.yaml');
  if (appValuesIndex > 0) return parts[appValuesIndex - 1];

  return 'unknown-app';
}

/**
 * Updates closed PR tracking references to point to 'main' branch.
 *
 * This function scans YAML configuration files for Git references that track
 * pull requests (in the format "pr-123") and checks if those PRs are closed.
 * If a PR is closed, the tracking reference is updated to "main".
 *
 * @param options - Configuration for the cleanup operation
 * @param options.contents - The YAML file contents to process
 * @param options.frozenEnvironments - Set of environment names that should not be modified
 * @param options.gitHubClient - GitHub client for API calls to check PR status
 * @param options.logger - Logger for operation feedback
 * @returns Promise resolving to updated contents and list of changes made
 */
export async function cleanupClosedPrTracking(options: {
  contents: string;
  frozenEnvironments: Set<string>;
  gitHubClient: GitHubClient;
  logger: PrefixingLogger;
  filename: string;
}): Promise<{ contents: string; changes: CleanupChange[] }> {
  const { contents, frozenEnvironments, gitHubClient, logger, filename } =
    options;

  const { document, stringify } = parseYAML(contents);
  if (!document) {
    return { contents, changes: [] };
  }

  const changes: CleanupChange[] = [];
  const trackables = findTrackables(document, frozenEnvironments);
  for (const trackable of trackables) {
    const match = trackable.trackMutableRef.match(/^pr-(\d+)$/);
    if (match && trackable.trackScalarTokenWriter) {
      const prNumber = parseInt(match[1], 10);
      try {
        const pr = await gitHubClient.getPullRequest({
          repoURL: trackable.repoURL,
          prNumber,
        });

        if (pr.state === 'closed') {
          trackable.trackScalarTokenWriter.write('main');
          logger.info(`PR #${prNumber} is closed, updated to main`);

          // Extract app name from filename and environment from blockKey
          const appName = extractAppNameFromFilename(filename);
          const environment = trackable.environment; // environment from YAML top-level key

          changes.push({
            prNumber,
            prTitle: pr.title,
            prURL: `${getWebURL(trackable.repoURL)}/pull/${prNumber}`,
            appName,
            environment,
            closedAt: pr.closedAt,
          });
        } else {
          logger.info(`PR #${prNumber} is ${pr.state}`);
        }
      } catch (error) {
        logger.info(`PR #${prNumber} lookup failed, leaving unchanged`);
      }
    }
  }

  return {
    contents: stringify(),
    changes,
  };
}
