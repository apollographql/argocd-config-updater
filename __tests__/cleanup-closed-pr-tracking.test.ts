import { cleanupClosedPrTracking } from '../src/update-closed-prs';
import { getWebURL, GitHubClient } from '../src/github';
import { PrefixingLogger } from '../src/log';

const logger = PrefixingLogger.silent();

function createMockGitHubClient(
  prStates: Record<number, 'open' | 'closed' | 'error'>,
): GitHubClient {
  return {
    async resolveRefToSHA() {
      return 'mock-sha';
    },
    async getTreeSHAForPath() {
      return 'mock-tree-sha';
    },
    async getCommitSHAsForPath() {
      return [];
    },
    async getPullRequest({ prNumber }) {
      const state = prStates[prNumber];
      if (state === 'error') {
        throw new Error('Not found');
      }
      return {
        state: state || 'open',
        title: `PR ${prNumber} title`,
        closedAt: state === 'closed' ? '2024-01-15T10:30:00Z' : null,
      };
    },
  };
}

describe('getWebURL', () => {
  it('should return a github web url given different repo url formats', () => {
    // Standard .git URLs
    expect(getWebURL('https://github.com/mdg-private/monorepo.git')).toBe(
      'https://github.com/mdg-private/monorepo',
    );
    expect(getWebURL('https://github.com/owner/repo.git')).toBe(
      'https://github.com/owner/repo',
    );

    // URLs without .git
    expect(getWebURL('https://github.com/owner/repo')).toBe(
      'https://github.com/owner/repo',
    );
    expect(getWebURL('https://github.com/mdg-private/monorepo')).toBe(
      'https://github.com/mdg-private/monorepo',
    );
  });
});

describe('cleanupClosedPrTracking', () => {
  it('should replace closed PR references with main', async () => {
    const contents = `
global:
  gitConfig:
    repoURL: https://github.com/owner/repo.git
    path: some/path
dev:
  track: pr-123
  gitConfig:
    ref: c0ffee
staging:
  track: main
  gitConfig:
    ref: c0ffee
prod:
  promote:
    from: staging
`;

    const gitHubClient = createMockGitHubClient({
      123: 'closed',
    });

    const result = await cleanupClosedPrTracking({
      contents,
      gitHubClient,
      logger,
      frozenEnvironments: new Set(),
      filename: 'teams/test-app/application-values.yaml',
    });

    expect(result.contents).not.toContain('track: pr-123');
    expect((result.contents.match(/track: main/g) || []).length).toBe(2);
    expect(result.contents).toMatchSnapshot();
    expect(result.changes).toHaveLength(1);
    expect(result.changes[0]).toMatchObject({
      prNumber: 123,
      prTitle: 'PR 123 title',
      prURL: 'https://github.com/owner/repo/pull/123',
      appName: 'test-app',
      environment: 'dev',
      closedAt: '2024-01-15T10:30:00Z',
    });
  });

  it('should not change pr references that are open', async () => {
    const contents = `
global:
  gitConfig:
    repoURL: https://github.com/owner/repo.git
    path: some/path
dev:
  track: pr-100
  gitConfig:
    ref: c0ffee
staging:
  track: pr-123
  gitConfig:
    ref: c0ffee
prod:
  promote:
    from: staging
`;

    const gitHubClient = createMockGitHubClient({
      123: 'closed',
      100: 'open',
    });

    const result = await cleanupClosedPrTracking({
      contents,
      gitHubClient,
      logger,
      frozenEnvironments: new Set(),
      filename: 'teams/test-app/application-values.yaml',
    });

    expect(result.contents).toContain('track: pr-100');
    expect(result.contents).not.toContain('track: pr-123');
    expect(result.contents).toContain('track: main');
    expect(result.changes).toHaveLength(1);
    expect(result.changes[0].prNumber).toBe(123);
    expect(result.contents).toMatchSnapshot();
  });

  it('should handle PR lookup errors by leaving PR references unchanged', async () => {
    const contents = `
global:
  gitConfig:
    repoURL: https://github.com/owner/repo.git
    path: some/path
dev:
  track: pr-999
  gitConfig:
    ref: c0ffee
`;

    const gitHubClient = createMockGitHubClient({
      999: 'error',
    });

    const result = await cleanupClosedPrTracking({
      contents,
      gitHubClient,
      logger,
      frozenEnvironments: new Set(),
      filename: 'teams/test-app/application-values.yaml',
    });

    expect(result.contents).toContain('track: pr-999');
    expect(result.contents).toMatchSnapshot();
    expect(result.changes).toHaveLength(0);
  });

  it('should not modify YAML unnecessarily', async () => {
    const contents = `# Top level comment
global:
  gitConfig:
    repoURL: https://github.com/owner/repo.git
    path: some/path
# Development environment
dev:
  track: pr-123  # Inline comment
  gitConfig:
    ref: c0ffee
staging:
  track: main  # Also tracking main
  gitConfig:
    ref: deadbeef
prod:
  promote:
    from: staging  # Promote from staging
`;

    const gitHubClient = createMockGitHubClient({
      123: 'closed',
    });

    const result = await cleanupClosedPrTracking({
      contents,
      gitHubClient,
      logger,
      frozenEnvironments: new Set(),
      filename: 'teams/test-app/application-values.yaml',
    });

    expect(result.contents).toContain('# Top level comment');
    expect(result.contents).toContain('# Inline comment');
    expect((result.contents.match(/track: main/g) || []).length).toBe(2);
    expect(result.contents).toMatchSnapshot();
    expect(result.changes).toHaveLength(1);
  });
});
