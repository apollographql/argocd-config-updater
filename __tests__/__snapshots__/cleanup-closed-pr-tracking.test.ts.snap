// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cleanupClosedPrTracking should handle PR lookup errors by leaving PR references unchanged 1`] = `
"
global:
  gitConfig:
    repoURL: https://github.com/owner/repo
    path: some/path
dev:
  track: pr-999
  gitConfig:
    ref: c0ffee
"
`;

exports[`cleanupClosedPrTracking should not change pr references that are open 1`] = `
"
global:
  gitConfig:
    repoURL: https://github.com/owner/repo
    path: some/path
dev:
  track: pr-100
  gitConfig:
    ref: c0ffee
staging:
  track: main
  gitConfig:
    ref: c0ffee
prod:
  promote:
    from: staging
"
`;

exports[`cleanupClosedPrTracking should not modify YAML unnecessarily 1`] = `
"# Top level comment
global:
  gitConfig:
    repoURL: https://github.com/owner/repo
    path: some/path
# Development environment
dev:
  track: main  # Inline comment
  gitConfig:
    ref: c0ffee
staging:
  track: main  # Also tracking main
  gitConfig:
    ref: deadbeef
prod:
  promote:
    from: staging  # Promote from staging
"
`;

exports[`cleanupClosedPrTracking should replace closed PR references with main 1`] = `
"
global:
  gitConfig:
    repoURL: https://github.com/owner/repo
    path: some/path
dev:
  track: main
  gitConfig:
    ref: c0ffee
staging:
  track: main
  gitConfig:
    ref: c0ffee
prod:
  promote:
    from: staging
"
`;
